x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java

```diff
@@ -479,7 +538,59 @@ public void accessGranted(String requestId, Authentication authentication, Strin
                         .withXForwardedFor(threadContext)
                         .with(authorizationInfo.asMap())
                         .build();
-                logger.info(AUDIT_MARKER, logEntry);
+            }
+        }
+        // "Security config change" records are not filtered out by ignore policies (i.e. they are always printed).
+        // The security changes here are the consequences of *user* requests,
+        // so in a strict interpretation we should filter them out if there are ignore policies in place for the causing user,
+        // but we do NOT do that because filtering out audit records of security changes can be unexpectedly dangerous.
+        if (events.contains(SECURITY_CONFIG_CHANGE) && SECURITY_CHANGE_ACTIONS.contains(action)) {
+            try {
```

Comment:
This approach bothers me.

Putting this code into the Audit trail rather than in the requests itself increases the risk that we will change the implementation of a request object, but fail to account for that in the audit logs (new fields, etc).
At one point we talked about an `AuditableRequest` interface with a `toAuditXContent` method. Is there  a reason we moved away from that idea?

Reply:
>At one point we talked about an AuditableRequest interface with a toAuditXContent method. Is there a reason we moved away from that idea?

I started off implementing that but changed course midway. I believe there are good reasons for the change of heart. Let me explain:

Conceptually the request bodies and the audit events both contain fields, and the code in this PR is all about the mapping between the two. The code has dependencies on both the accessor methods of requests (and, optionally, on the toXContent method, which is used when indexing the object) and the "setter" methods for the audit event. After some prototyping, I've reached the conclusion that the "mapping" code is more dependent on the auditing part than on the request content part. The nestedness and the field names (some of which must be made all the way to the log4j2.properties file) are all examples of it; they are like this because they're crafted to present a consistent interface for all the security changes (eg `"user":{"name"` vs `"username"), which is the being reason for this feature (the "upgrade" from the REST request auditing). It is useful to have all the formatting code in a single place. Consequently, the code sits in the `LoggingAuditTrail` class rather than the individual request classes.

> Putting this code into the Audit trail rather than in the requests itself increases the risk that we will change the implementation of a request object, but fail to account for that in the audit logs (new fields, etc).

I acknowledge the worry, but the `LoggingAuditTrail` is now a consumer of such requests just like the action handlers; if you change the request you can track all the uses of the request class in the IDE and adapt all the consumers as well (not only the conventional consumer which is the action handler). It's not fool proof, but neither is a "toString"-like method, which you can miss to adapt when you change the class.

Reply:
I've reread the comment and I think I came across defensive. I've given this PR a lot of attention and to the best of my judgement this is the best option of the two.
But we can change the approach no problem, it will only be a small amount of work.

Reply:
> you can track all the uses of the request class in the IDE and adapt all the consumers as well

I think we need something more reliable than that, but I'm happy to add it in a followup.
I'm thinking something like a test in the audit package that verifies that the request/request builder matches the format we expect, so that changes in the request, will force a matching change in the audit implementation.

Reply:
We've discussed this on a different channel.

Whenever a request body changes, it is expected that audit entries change as well, eg adding a new request field.
BUT, it's not easy to figure out if the new field should indeed be audited or under what field name. It's not alike request serialization where every field must be serialized. For example, a confidential field might be audited only as present or not, or audited under a slightly different name for uniformity in the audit context.

Therefore, there's no advantage to placing the auditing code at the request level. **We must ensure that changes to request objects of security actions are reflected in all the usage sites of those requests, in this case the auditing code in addition to the action handling code.**